text/0000-language-semver.md

```diff
@@ -0,0 +1,494 @@
+  temporary deprecation, that can be used to ease the transition
+  around a soundness fix. Moreover, separating out new type rules so
+  that they can be "opted into" can be very difficult and would
+  complicate the compiler internally; it would also make it harder to
+  reason about the type system as a whole.
+
+**Rather than using a version number to opt-in to minor changes, one
+might consider using the existing feature mechanism.** For example,
+one could write `#![feature(foo)]` to opt in to the feature "foo" and
+its associated keywords and type rules, rather than
+`#![rust_version="1.2.3"]`. While using minimum version numbers is
+more opaque than named features, they do offer several advantages:
+
+1. Using a version number alone makes it easy to think about what
+   version of Rust you are using as a conceptual unit, rather than
+   choosing features "a la carte".
+2. Using named features, the list of features that must be attached to
+   Rust code will grow indefinitely, presuming your crate wants to
+   stay up to date.
+3. Using a version attribute preserves a mental separation between
+   "experimental work" (feature gates) and stable, new features.
+4. Named features present a combinatoric testing problem, where we
+   should (in principle) test for all possible combinations of
+   features.
+   
+# Unresolved questions
+
+**Can (and should) we give a more precise definition for compiler bugs
+and soundness problems?** The current text is vague on what precisely
+constitutes a compiler bug and soundness change. It may be worth
+defining more precisely, though likely this would be best done as part
+of writing up a more thorough (and authoritative) Rust reference
+manual.
+
+**Should we add a mechanism for "escaping" keywords?"** We may need a
+mechanism for escaping keywords in the future. Imagine you have a
+public function named `foo`, and we add a keyword `foo`. Now, if you
+opt in to the newer version of Rust, your function declaration is
+illegal: but if you rename the function `foo`, you are making a
+breaking change for your clients, which you may not wish to do. If we
+had an escaping mechanism, you would probably still want to deprecate
+`foo` in favor of a new function `bar` (since typing `foo` would be
+awkward), but it could still exist.
+
+**Should we add a mechanism for skipping over new syntax?** The
+current `#[cfg]` mechanism is applied *after* parsing. This implies
+that if we add new syntax, crates which employ that new syntax will
+not be parsable by older compilers, even if the modules that depend on
+that new syntax are disabled via `#[cfg]` directives. It may be useful
+to add some mechanism for informing the parser that it should skip
+over sections of the input (presumably based on token trees). One
+approach to this might just be modifying the existing `#[cfg]`
+directives so that they are applied during parsing rather than as a
+post-pass.
```

Comment:
@pnkfelix 

> As a potentially less invasive form of this idea...[just skip modules]

Seems like a good starting point. I was originally thinking we'd just skip over a token tree's worth of syntax to represent an item, but of course that doesn't work. It would if we had some sort of 'parentheses' for items:

```
#[cfg(...)]
{ // no semantics, just to group
    struct Foo { }
}
```

but of course that's not how it works.